src/date_extraction.py

```diff
@@ -1,17 +1,34 @@
 import re
+import datetime as dt
 
 delimiter = "(\.|_| |-|\s)"
-#pattern_1 finds day-month-year with delimiters
-pattern_1 = "([1-9]|0[1-9]|1[\d]|2[\d]|3[0-1])"+delimiter+"([1-9]|0[1-9]|1[0-2])"+delimiter+"([\d][\d]|19[\d][\d]|20[\d][\d])"
-#pattern_2 finds year-month-day patterns with delimiters
-pattern_2 = "([\d][\d]|19[\d][\d]|20[\d][\d])"+delimiter+"([1-9]|0[1-9]|1[0-2])"+delimiter+"([1-9]|0[1-9]|1[\d]|2[\d]|3[0-1])"
-#pattern 3 finds strings of eight consecutive digits that match a year-month-day format
-pattern_3 = "(19[\d][\d]|20[\d][\d])(0[1-9]|1[0-2])(0[1-9]|1[\d]|2[\d]|3[0-1])"
-#pattern 4 finds strings of eight consecutive digits that match a day-month-year format
-pattern_4 = "(0[1-9]|1[\d]|2[\d]|3[0-1])(0[1-9]|1[0-2])(19[\d][\d]|20[\d][\d])"
-dates_regex_list  = [pattern_1, pattern_2, pattern_3, pattern_4]
+patterns_list = [{"pattern": "([1-9]|0[1-9]|1[\d]|2[\d]|3[0-1])"+delimiter+"([1-9]|0[1-9]|1[0-2])"+delimiter+"(19[\d][\d]|20[\d][\d])",
+                  "strp_format": "%d/%m/%Y"},
+                 {"pattern":"(19[\d][\d]|20[\d][\d])"+delimiter+"([1-9]|0[1-9]|1[0-2])"+delimiter+"([1-9]|0[1-9]|1[\d]|2[\d]|3[0-1])",
+                  "strp_format": "%Y/%m/%d"},
+                 {"pattern": "(19[\d][\d]|20[\d][\d])(0[1-9]|1[0-2])(0[1-9]|1[\d]|2[\d]|3[0-1])",
+                  "strp_format": "%Y%m%d"},
+                 {"pattern": "(0[1-9]|1[\d]|2[\d]|3[0-1])(0[1-9]|1[0-2])(19[\d][\d]|20[\d][\d])",
+                  "strp_format": "%d%m%Y"}
+                 ]
```

Comment:
would be good to keep comments if possible

Reply:
Also I'm kind of wary of handling d/m/Y, just because it's indistinguishable from m/d/Y.

Reply:
Not sure how extensively you've tested these patterns but I could imagine they might not always work as intended. For example it seems 1_1_1999 could be a valid match for the first pattern, but it won't match the ["%d/%m/%Y" format as d and m are 2-digit numbers](https://devhints.io/strftime). 

Additionally I'm not a fan of restricting the range of possible values, for example year must be 19xx or 20xx. I know it's niche but there could be some reason for dates outside this range in filenames and restricting seems somewhat arbitrary. I think it would be better to look for sequences of any number of digits of the right length and potentially use a try/except when parsing into a date to handle any non-dates that slip through.

Reply:
Thanks Chris, have kept the regex simple instead of narrowing down the options from the start. Only a few very general patterns accepted. Included a try, except to raise value errors in parser arguments. This could be further improved to deal with general problems but seems good as a backbone.